addOns/websocket/src/main/java/org/zaproxy/zap/extension/websocket/pscan/WebSocketPassiveScanThread.java

```diff
@@ -0,0 +1,182 @@
+     * alive. In order to activate thread set {@link
+     * WebSocketPassiveScanThread#setPassiveScannerActive(boolean)} true and start thread with
+     * {@link WebSocketPassiveScanThread#start()}.
+     *
+     * @param passiveScannerManager the manager
+     */
+    public WebSocketPassiveScanThread(WebSocketPassiveScannerManager passiveScannerManager) {
+        this.passiveScannerManager = passiveScannerManager;
+        this.isPassiveScannerActive = false;
+        messagesBuffer = new LinkedBlockingQueue<>();
+    }
+
+    /** @return true if the table was initialized */
+    public boolean hasTableWebSocket() {
+        return tableWebSocket != null;
+    }
+
+    public void setTableWebSocket(TableWebSocket tableWebSocket) {
+        this.tableWebSocket = tableWebSocket;
+    }
+
+    public boolean isPassiveScannerActive() {
+        return isPassiveScannerActive;
+    }
+
+    public void setPassiveScannerActive(boolean passiveScannerActive) {
+        isPassiveScannerActive = passiveScannerActive;
+    }
+
+    @Override
+    public int getObservingOrder() {
+        return WEBSOCKET_OBSERVING_ORDER;
+    }
+
+    @Override
+    public boolean onMessageFrame(int channelId, WebSocketMessage message) {
+        if (message.isFinished() && !shouldIgnoreServerModeMessages(message)) {
+            messagesBuffer.add(new MessageWrapper(message.getMessageId(), channelId));
+        }
+        return true;
+    }
+
+    @Override
+    public void onStateChange(WebSocketProxy.State state, WebSocketProxy proxy) {
+        // Ignore
+    }
+
+    @Override
+    public void run() {
+        MessageWrapper messageWrap;
+        WebSocketPassiveScannerPlugin currentPassiveScanner;
+        WebSocketMessageDTO currentMessage;
+        Iterator<WebSocketPassiveScanner> iterator;
+
+        while (isPassiveScannerActive) {
+            if (messagesBuffer.isEmpty() || tableWebSocket == null) {
+                try {
+                    Thread.sleep(SLEEP_TIME);
+                } catch (InterruptedException e) {
+                    LOGGER.info("Sleeping was interrupted", e);
+                }
+                if (!isPassiveScannerActive) {
+                    break;
+                }
+            } else {
+                messageWrap = messagesBuffer.poll();
+                try {
+                    currentMessage =
+                            tableWebSocket.getMessage(messageWrap.messageId, messageWrap.channelId);
+                    iterator = passiveScannerManager.getIterator();
+                    while (iterator.hasNext()) {
+                        currentPassiveScanner = (WebSocketPassiveScannerPlugin) iterator.next();
+                        if (currentPassiveScanner.isEnabled()) {
+                            currentPassiveScanner.scanMessage(
+                                    new WebSocketScanHelper(this, currentPassiveScanner.getId()),
```

Comment:
The `WebSocketScanHelper` could be reused for all scanners, it needs some tweaks though (e.g. allow to set the ID).

Reply:
Fixed in 7abd324